architecture/2025-01-serverless-feature-spec.md

```diff
@@ -0,0 +1,279 @@
+    }
++    providers: { // This was introduced as a part of the ACI prototype, but may be changed depending on the implementation
++      azure: {
++        scope: '/subscriptions/.../resourceGroups/myrg'
++      }
++    }
+    extensions: [
+      {
+        kind: 'kubernetesMetadata'
+        labels: {
+          'team.contact.name': 'frontend'
+        }
++        // Add other serverless platform-specific extensions here
+      }
+    ]
+  }
+}
+```
+
+### Step 2: Define a Radius Application for serverless
+
+The user defines a new Radius Application for serverless compute by creating a new Application definition file (e.g. `app.bicep`) and specifying the necessary settings for the container runtime platform. Note that the application definitions are container runtime platform-agnostic, thus this same application definition can be deployed to both Kubernetes and serverless compute platforms.
+
+```diff
+resource app 'Applications.Core/applications@2023-10-01-preview' = {
+  name: 'myapp'
+  properties: {
+    environment: environment
++    extensions: [ // these are all currently Kubernetes-centric, will need to update for serverless
+      {
+        kind: 'kubernetesNamespace'
+        namespace: 'myapp'
+      }
+      {
+        kind: 'kubernetesMetadata'
+        labels: {
+          'team.contact.name': 'frontend'
+        }
+      }
++      // add other serverless extensions as applicable
+    ]
+  }
+}
+```
+
+### Step 3: Define a Radius Container for serverless
+
+The user defines a Radius Container within the application definition (e.g. `app.bicep`) and specifies relevant container properties, such as `extensions` or `runtimes` to set platform specific configurations. Note that the container definitions are container runtime platform-agnostic, thus this same container definition can be deployed to both Kubernetes and serverless compute platforms if common functionalities across compute platforms are used.
+
+```diff
+resource demo 'Applications.Core/containers@2023-10-01-preview' = {
+  name: 'demo'
+  properties: {
+    application: application
+    container: {
+      image: 'ghcr.io/radius-project/samples/demo:latest'
+      ports: {
+        web: {
+          containerPort: 3000
+        }
+      }
++      // all the other container properties should also be implemented (e.g. `env`, `readinessProbe`, `livenessProbe`, etc.)
+    }
+    extensions: [
++      {
++        kind:  'manualScaling'
++        replicas: 2
++      }
+    ]
+    runtimes: {
++      aci: {
++        // Add ACI-specific properties here to punch-through the Radius abstraction, e.g. sku, osType, etc.
++        sku: 'Confidential' // 'Standard', 'Dedicated', etc.
++        }
++      ecs: {
++        // Add AWS ECS-specific properties here to punch-through the Radius abstraction
++        }
++      }
```

Comment:
Is the expectation here that only one runtime will be allowed at once and it has to match what's configured as the compute for the environment?

Reply:
To make the app definition portable, we must allow several `runtimes` to be declared per container and these should all be optional for cases where the user wants to punch through the Radius abstraction. If they declare a `runtimes` property that doesn't match the targeted deployment environment's compute, we should simply ignore that property for that deployment. If they haven't declared any `runtimes` property that match the compute of the targeted deployment environment, then we should deploy their container assuming that no `runtimes` property was provided and thus no "punch-through" behavior will be applied.

I'll add this as a note to the spec, good question.